### ISSUE TITLE
managing virtual env and npm packages from within jupyter lab

### ISSUE THREAD

**Issue by rafikam** [jupyterlab/jupyterlab#3258]
Entirely new to jupyter here... If I wanted to give capability for users to create, manage, and select virtual envs for them to work in from within the jupyter lab UI, is that something that should be theoretically possible from inside a plugin? I'm assuming plugins would be a better fit for environment setup/management than an extension.

I see advice out there for setting up different kernels outside of jupyter and then toggling between them through built-in menu but I'm looking to give the ability to do venv and package management from within jupyterlab. Much appreciated if you could point me in the right direction.

Apologies for posting this kind of request here but there's very little on jupyterlab on stackoverflow. Let me know if somewhere more suitable to ask. Thanks in advance.

**Comment by blink1073** [comment]
Hi @rafikam, I think that would involve creating a [server extension](http://jupyter-notebook.readthedocs.io/en/stable/extending/handlers.html#Writing-a-notebook-server-extension) that includes a custom [ContentsManager](http://jupyter-notebook.readthedocs.io/en/latest/extending/contents.html#writing-a-custom-contentsmanager) as well as a [KernelSpecManager](https://github.com/Anaconda-Platform/nb_conda_kernels/blob/f7b468090fc2f7fbf1dd56fe875f3e5c28051aa9/nb_conda_kernels/manager.py#L20).

**Comment by rafikam** [comment]
Thanks! That's very helpful.

So if I was looking do this from within jupyterlab, would I need to update the jupyterlab server extension itself? Or are separate server extensions something that can be utilized from jupyterlab?

**Comment by blink1073** [comment]
You'd need both parts: the server extension to handle the requests from the frontend, and a frontend extension to allow the user to configure it.

**Comment by rafikam** [comment]
I guess what I'm asking is are notebook server extensions a concept within jupyterlab as well? I don't see it in those docs, aside from describing jupyterlab itself as a server extension. So confirming that's what I'd be looking to customize? In addition to creating the front-end extension...

**Comment by blink1073** [comment]
Yes, we use the same server.  Eventually we will break out a separate `jupyter-server` and the classic notebook will become a server extension itself.  

Here are our extension developer docs:
http://jupyterlab-tutorial.readthedocs.io/en/latest/extensions_dev.html
http://jupyterlab-tutorial.readthedocs.io/en/latest/xkcd_extension_tutorial.html

**Comment by ZainRizvi** [comment]
@rafikam Did you end up developing that extension?  If you did, I'd love a pointer to it :)
